src/generic/Matrix.jl

```diff
@@ -2371,6 +2323,104 @@ function can_solve_left_reduced_triu(r::AbstractAlgebra.MatElem{T},
    return true, x
 end
 
+@doc Markdown.doc"""
+    can_solve_with_solution(a::AbstractAlgebra.MatElem{S}, b::AbstractAlgebra.MatElem{S}; side = :right) where S <: RingElement
+> Given two matrices $a$ and $b$ over the same ring, tries to solve $ax = b$
+> if `side` is `:right` or $xa = b$ if `side` is `:left`. In either case,
+> returns a tuple `(flag, x)`. If a solution exists, `flag` is set to true and
+> `x` is a solution. If no solution exists, `flag` is set to false and `x`
+> is arbitrary. If the dimensions of $a$ and $b$ are incompatible, an exception
+> is raised.
+"""
+function can_solve_with_solution(a::AbstractAlgebra.MatElem{S}, b::AbstractAlgebra.MatElem{S};
+                                                                                side = :right) where S <: RingElement
+   if side == :right
+      (f, x) = can_solve_with_solution(a', b'; side=:left)
+      return (f, x')
+   elseif side == :left
+      @assert ncols(a) == ncols(b)
+      H, T = hnf_with_transform(a)
+      b = deepcopy(b)
+      z = zero(a, nrows(b), nrows(a))
+      l = min(ncols(a), nrows(a))
+      t = base_ring(a)()
+      for i = 1:nrows(b)
+         for j = 1:l
+            k = 1
+            while k <= ncols(H) && iszero(H[j, k])
+               k += 1
+            end
+            if k > ncols(H)
+               continue
+            end
+            q, r = divrem(b[i, k], H[j, k])
+            if r != 0
+               return (false, zero(a, 0, 0))
+            end
+            z[i, j] = q
+            q = -q
+            for h = k:ncols(H)
+               t = mul!(t, q, H[j, h])
+               b[i, h] = addeq!(b[i, h], t)
+            end
+         end
+      end
+      if b != 0
+         return (false, zero(a, 0, 0))
```

Comment:
> Where do you mean? As far as I can tell the variables are all necessary,

@hwjsnc I meant re-using the input matrix, like `a` here for example, instead of `zero(a, 0, 0)`, to avoid allocating a dummy matrix.